### DISCUSSION TITLE
Update a model from inside a command handler

### DISCUSSION THREAD

**Discussion by luiz290788** [eclipse-langium/langium#1029] [Q&A] [answered]
Is there any way to update a document from inside of a command handler? Do I need to trigger another command?

**Comment by msujew** [comment]
Hey @luiz290788,

what exactly are you trying to accomplish? In what ways are you looking to update the document? Do you want to [trigger a code action](https://code.visualstudio.com/docs/editor/refactoring) programmatically, so that the editor value is actually changed? Or is it enough to just update the document internally in Langium?

**Reply by luiz290788** [reply]
My use case, I use a language to generate code for another language. I was bringing AST to the main thread and then updating the model using monaco. But because of [this](https://github.com/langium/langium/discussions/1028), I was facing some problems serializing the AST to the main thread, building the code for the other language and setting the value in the model.

Also if I can do everything in the language server, I think that the interface will be faster without adding extra overhead to the ui thread.

**Reply by msujew** [reply]
I see. Yeah, it would absolutely make sense to run the code generator that you have inside of the language server. Commands can return arbitrary data after all, so you can just build a `generate` command that returns the whole generated code for a document/workspace as a string. The ui thread can just await the result of the command and then set the result as the new editor value. This also has the added benefit of not blocking the ui thread during code generation.

**Reply by luiz290788** [reply]
That's what we are doing right now, but we were wondering if we need to send the generated content back to the ui thread. Is there any way that we can set the value of the document from inside of the language server?

**Reply by msujew** [reply]
You could always just generate a document from the string directly via `LangiumDocumentFactory#fromString`, but then the user won't be able to see the document. Language servers work in a master-slave architecture, where the editor has complete control over the text document and only consults the language server for certain requests. The editor will never ask the server what the state of the document is, it will only push changes to the server.

**Reply by luiz290788** [reply]
I see, so the best way is to send the generated content back to the main thread and let the editor save the content.

**Reply by msujew** [reply, marked as answer]
I think so. Langium's document representation is completely internalized and isolated from "the real world". There's no way to just modify the document and have the changes reflected anywhere else, since they're decoupled.
